---
title: "Look Up"
subtitle: ""
author: 
- name: Morteza Sabri
  affiliation: 
    Institute of Molecular Vascular Medicine, Klinikum rechts der Isar, Technical University of Munich
  email: morteza.sabri@tum.de
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc : true
    toc_depth : 2
    df_print: paged
    code_folding: 'show'
    toc_float: false
    keep_md: true
params: 
  goi: "GOI"
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r inputs, include=FALSE}
dir <- c("~/Workspace/projects/lars_aaa2021",
         "~/Workspace/projects/lars_plaques2022", 
         "~/Workspace/projects/lars_plaques2023")
dir <- dir[1]           # select the dir
# goi <- "MT2A"           # put a gene
# goi <- function(goi = readline("Enter name of the gene (goi)")) {
#   return(goi)
# }

readname = function()# Get the project name
{ 
  params$goi
}
goi=readname()
goi
```


```{r setup, include=FALSE, message=FALSE, warning=FALSE}
list.dirs.depth.n <- function(p, n) {
  res <- list.dirs(p, recursive = FALSE)
  if (n > 1) {
    add <- list.dirs.depth.n(res, n-1)
    c(res, add)
  } else {
    res
  }
}

options(width = 100, timeout = 600)
knitr::opts_chunk$set(echo = FALSE, dpi = 200, message = FALSE, warning = FALSE)


interestedTerms <- c("cardi", "arter", "vascul", "inflammat")

tmp <- getwd()
setwd(dir)

project <- list.dirs.depth.n("..", 3)
project <- grep("out/salmon", project, value = TRUE)
project <- basename(project)
project <- grep(sub("lars_", "", basename(dir)), 
                project, value = TRUE, ignore.case = TRUE)
idat <- paste0( "idat/", project, "/")

load(paste0(idat, "de.RData")) # the RData object which save before from DE analysis

setwd(tmp)
outdir <- paste0(project, goi, "/")
if(!dir.exists(outdir)) dir.create(outdir, recursive = TRUE)
```

# Summary

```{r abstract, include=FALSE}
library(biomaRt)
# mart <- readRDS("~/Workspace/dat/rnaseq/hsa_mart.rds")
des <- biomaRt::getBM(attributes=c(rowNamesOfCounts, "external_gene_name",
                          'ensembl_gene_id', 
                          'gene_biotype', "description", 
                          'chromosome_name', 
                          'start_position', 'end_position', 'strand'), 
      filters = rowNamesOfCounts, 
      values = goi, 
      mart = mart)
entrezgene_id <- AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db, 
                                       keys = goi, 
                                       keytype = "SYMBOL", column="ENTREZID")

link <- unique(paste0("https://useast.ensembl.org/", 
       sub(" ", "_", ScientificName),
       "/Gene/Summary?g=", des$ensembl_gene_id))
```


**Gene summary:** `r goi`

- Gene Symbol: [`r unique(des$external_gene_name)`](`r link`)
- Ensembl ID: `r unique(des$ensembl_gene_id)`
- Entrez (NCBI Gene) ID:  `r entrezgene_id`
- Gene Biotype: `r unique(des$gene_biotype)`
- Description: `r as.name(unique(des$description))`
- [Gene Ontologies](`r sub("Summary", "Ontologies/molecular_function", link)`)
- [Orthologues](`r sub("Summary", "Compara_Ortholog", link)`)
- [Pathways](`r sub("Summary", "Pathway", link)`)

All the plots and comparisons are based on this project (dataset):

- `r project`
- P-value of `r goi` in this project: `r format(res[goi, "padj"], digits = 2)` 
- LogFC of `r goi` in this project: `r round(res[goi, "log2FoldChange"], 1)` 

Study design:

- **Comparison**: `r DESeq2::resultsNames(dds)[2]`
- **Organism**: `r organism`
- **Number of samples**: **`r ncol(dds)`** (Healthy: `r length(healthy)`, Disease: `r length(case)`)
- **Excluded Samples**: `r ifelse(exists("excludedSample"), excludedSample, "None")`
- **Control samples**: `r colnames(counts)[grepl(abr_healthy, colnames(counts))]`
- **Case samples**: `r colnames(counts)[grepl(abr_case, colnames(counts))]`


```{r}
df <- geneCounts <-
    DESeq2::plotCounts(
      dds,
      gene = goi,
      intgroup = "condition",
      returnData = TRUE)
```


# Barplot

```{r Barplot}
library(ggpubr)
library(rstatix)
palette <- c("#4f8832", "#f79c18")

stat.test <- df %>% t_test(count ~ condition)
stat.test$p <- round(res.df[goi, ]$padj, 3)
stat.test$statistic <- round(res.df[goi, ]$stat, 3)
stat.test <- add_significance(stat.test)
stat.test <- stat.test %>% add_xy_position(fun = "mean_sd", x = "condition")

g <- ggbarplot(df, x = "condition", y = "count", add = "mean_sd", 
               fill = "condition", 
                palette = palette) + 
  stat_pvalue_manual(stat.test, label = "padj: {p} {p.signif}", 
                     tip.length = 0.01) +
  labs(title = paste0("Barplot for ", goi), y = "Normalized counts")
g
ggsave(paste0(outdir, "barplot_", goi, ".png"),
         g,
         dpi = 300)
```


# Boxplot


```{r Boxplot}
stat.test <- df %>% t_test(count ~ condition); 
stat.test$p <- round(res.df[goi, ]$padj, 3)
stat.test$statistic <- round(res.df[goi, ]$stat, 3)
stat.test <- add_significance(stat.test)
stat.test <- stat.test %>% add_xy_position(x = "condition")

rm(g); g <- ggboxplot(df, x = "condition", y = "count", 
                 fill = "condition", 
                 palette = palette) +
  geom_point(position=position_dodge(width=0.75), 
             aes(group=condition), 
             color = "steelblue4") + 
  stat_pvalue_manual(stat.test, label = "padj: {p} {p.signif}",
  vjust = -1, bracket.nudge.y = 1) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.15)))  + 
  labs(title = paste0("Boxplot for ", goi), y = "Normalized counts")
g
ggsave(paste0(outdir, "boxplot_", goi, ".png"),
         g,
         dpi = 300)
```


# Count Plot

```{r Count Plot}
if(!requireNamespace("ggbeeswarm")) install.packages("ggbeeswarm")
library("ggbeeswarm")

geneCounts <-
    DESeq2::plotCounts(
      dds,
      gene = goi,
      intgroup = "condition",
      returnData = TRUE)

rm(g); g <- ggplot(geneCounts, aes(x = condition, y = count, color = condition)) +
    scale_y_log10() + geom_beeswarm(cex = 3) + ggtitle(goi) +
    theme(legend.position = "none")
g
ggsave(paste0(outdir, "plotCounts_", goi, ".png"),
         g,
         dpi = 300)
```


# Volcano Plot

```{r Volcano}
library(ggrepel)

res.df$threshold <- res.df$padj < pCutoff & abs(res.df$log2FoldChange) > lfcCutoff
res.df$genelabels <- rownames(res.df) %in% goi # name of genes want to display
options(ggrepel.max.overlaps = Inf) # in case of ggrepel: 1 unlabeled data points (too many overlaps). Consider increasing max.overlaps


rm(g); g <- ggplot(res.df) +
  geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold), 
             size =0.4) +
  geom_text_repel(aes(x = log2FoldChange, y = -log10(padj), 
                      label = ifelse(genelabels == T, 
                                     rownames(res.df),""))) +
  ggtitle( paste0("Volcano Plot (", goi, ")")) +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) 
g
ggsave(paste0(outdir, "volcanoPlot_", goi, ".png"),
         g,
         dpi = 300)
```



# MA Plot 

```{r MA plot, echo=FALSE}
library(DESeq2)

plotMA(res, ylim = c(-4,4), main = "MA Plot")
with(subset(res, rownames(res) %in% goi), {
  text(baseMean, log2FoldChange, goi, pos = 2, col = "black", cex = 1) 
  })
```


```{r MA plot2, include=FALSE, echo=FALSE}
png(paste0(outdir, "MAplot_", goi, ".png"), 
    units = "in", height = 10, width = 14, res = 200)
plotMA(res, ylim = c(-4,4), main = "MA Plot")
with(subset(res, rownames(res) %in% goi), {
  text(baseMean, log2FoldChange, goi, pos = 2, col = "black", cex = 1) 
  })
dev.off()
```


# Association with Diseases

```{r}
x <- data.frame(KnowSeq::DEGsToDiseases(goi, size = 10, getEvidences = TRUE))[, 1:2]
x[,2] <- as.numeric(x[,2])
x[,2] <- round(x[,2], 3)
knitr::kable(x)
```

More info on the [Open Targets](`r paste0("https://platform.opentargets.org/target/", unique(des$ensembl_gene_id) ,"/associations")`)

# Pathways which `r goi` is included

```{r Pathways}
outKEGG <- paste0(outdir, "KEGG/")
if (!dir.exists(outKEGG)) dir.create(outKEGG, recursive = TRUE)
tmp <- getwd()
setwd(outKEGG)

geneList <- degs$log2FoldChange; names(geneList) <- degs$entrezgene_id
geneList <- geneList[complete.cases(names(geneList))]


z <- limma::getGeneKEGGLinks("hsa")
path.ids <- z[z$GeneID == entrezgene_id, ]$PathwayID

if (!requireNamespace("KEGGREST")) BiocManager::install("KEGGREST")
library(KEGGREST)
for (i in 1:length(path.ids)) {
  names(path.ids)[i] <- keggGet(path.ids[i])[[1]]$PATHWAY
}

if (!requireNamespace("pathview")) BiocManager::install("pathview")
library(pathview)
for (i in 1:length(path.ids)) {
  tryCatch({
      
  var2 <- path.ids[i] #b[i]
  nm <- paste0(var2, "_", names(var2))
  message(paste0("\n *** Initiating loop ", i, ": ", nm))
    
  geneOnThePathID <- keggGet(path.ids[i])[[1]]$GENE
  geneOnThePathID <- geneOnThePathID[seq(1, length(geneOnThePathID), by = 2)]
  geneOnThePathID <- geneList[which(names(geneList) %in% geneOnThePathID)]
  ceil <- max(abs(geneOnThePathID))
  

    pathview::pathview(
      gene.data = geneList,
      pathway.id = path.ids[i],
      species = abrScientificName,
      out.suffix = names(var2),
      kegg.native = T,
      limit = list(gene = ceil, cpd = 1)
    )
  }, error = function(e) {
  })
}

dir.create("paths") 
filesstrings::file.move(list.files(pattern = paste0(abrScientificName,
                                                    "[0-9]*\\..*\\.png")), 
                        "paths") # hsa for human
setwd(tmp)
```


```{r}
l <- list.files(file.path(outKEGG, "paths"), full.names = T)

knitr::include_graphics(l)
```



# GWAS (NOT related to local datasets)

```{r}
x <- otargen::colocalisationsForGene("TP53")
x <- as.data.frame(x)
if (nrow(x) > 0) {
  if (length(x) > 10) {
    ind <- grep(paste(interestedTerms, collapse = "|") , 
            x$Trait_reported, ignore.case = TRUE)
    message("Here are the most interested ones")
    knitr::kable(x[ind, ])
  } else {
    message("Here are the all traits associated with this gene")
    knitr::kable(x)
  }
}
```



```{r}
otargen::plot_manhattan(otargen::manhattan(x$Study[2]))
```


# Session info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
